util/failuredomains/failure_domains_test.go

```diff
@@ -172,12 +179,696 @@ func TestNewFailureDomainPickMost(t *testing.T) {
+			fds:              fds0,
+			allMachines:      collections.FromMachines(machineA1, machineB1, machineC1),
+			eligibleMachines: collections.FromMachines(),
+			expected:         nil,
+		},
+		{
+			name:             "No machines", // Note: this should never happen (scale down is called only when there are machines)
+			fds:              fds0,
+			allMachines:      collections.FromMachines(),
+			eligibleMachines: collections.FromMachines(),
+			expected:         nil,
+		},
+	}
+	for _, tc := range testcases {
+		t.Run(tc.name, func(t *testing.T) {
+			g := NewWithT(t)
+
+			fd := PickMost(ctx, tc.fds, tc.allMachines, tc.eligibleMachines)
+			if tc.expected == nil {
+				g.Expect(fd).To(BeNil())
+			} else {
+				g.Expect(fd).ToNot(BeNil())
+				g.Expect(tc.expected).To(ContainElement(*fd))
 			}
 		})
 	}
 }
+
+func TestCountByFailureDomain(t *testing.T) {
+	g := NewWithT(t)
+
+	a := "us-west-1a"
+	b := "us-west-1b"
+
+	fds := clusterv1.FailureDomains{
+		a: clusterv1.FailureDomainSpec{ControlPlane: true},
+		b: clusterv1.FailureDomainSpec{ControlPlane: true},
+	}
+	machinea1 := &clusterv1.Machine{ObjectMeta: metav1.ObjectMeta{Name: "a1"}, Spec: clusterv1.MachineSpec{FailureDomain: ptr.To(a)}}
+	machinea2 := &clusterv1.Machine{ObjectMeta: metav1.ObjectMeta{Name: "a2"}, Spec: clusterv1.MachineSpec{FailureDomain: ptr.To(a)}}
+	machineb1 := &clusterv1.Machine{ObjectMeta: metav1.ObjectMeta{Name: "b1"}, Spec: clusterv1.MachineSpec{FailureDomain: ptr.To(b)}}
+	machinenil := &clusterv1.Machine{ObjectMeta: metav1.ObjectMeta{Name: "nil"}, Spec: clusterv1.MachineSpec{FailureDomain: nil}}
+
+	allMachines := collections.FromMachines(machinea1, machinea2, machineb1, machinenil)
+	priorityMachines := collections.FromMachines(machinea1)
+
+	aggregations := countByFailureDomain(ctx, fds, allMachines, priorityMachines)
+
+	g.Expect(aggregations).To(HaveLen(2))
+	g.Expect(aggregations).To(ConsistOf(failureDomainAggregation{id: a, countPriority: 1, countAll: 2}, failureDomainAggregation{id: b, countPriority: 0, countAll: 1}))
+}
+
+func TestFailureDomainAggregationsSort(t *testing.T) {
+	g := NewWithT(t)
+
+	aggregations := failureDomainAggregations{
+		{id: "fd1", countPriority: 2, countAll: 2},
+		{id: "fd2", countPriority: 2, countAll: 3},
+		{id: "fd3", countPriority: 2, countAll: 2},
+		{id: "fd4", countPriority: 1, countAll: 5},
+	}
+
+	sort.Sort(aggregations)
+
+	// the result should be sorted so the fd with less priority machines are first,
+	// the number of overall machine should be used if the number of priority machines is equal,
+	// failure domain id should be used to break ties.
+
+	// fd4 is the failure domain with less priority machines, it should go first
+	g.Expect(aggregations[0].id).To(Equal("fd4"))
+
+	// fd1, fd2, fd3 all have the same number of priority machines;
+
+	// fd1, fd3 have also the same number of overall machines.
```

Comment:
I guess this test is actually determenistic because the input array is always the same and we preserve order?

So the or's below are not required